src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -35,6 +37,7 @@ public abstract class GeneratorDriver
 
         internal GeneratorDriver(GeneratorDriverState state)
         {
+            Debug.Assert(state.Generators.GroupBy(s => s.GetType()).Count() == state.Generators.Length); // ensure we don't have duplicate generator types
```

Comment:
Why do you want to restrict generator types to be unique? Consider the case where someone authors a really handy generator library and two different generators consume it. That seems like a valid case where we would have duplicate generator types.

Reply:
Although writing that I realize that the full name of the type isn't actually enough to be unique right? I can have two types with the same full name in two different assemblies?

---
In reply to: [434750141](https://github.com/dotnet/roslyn/pull/44768#discussion_r434750141) [](ancestors = 434750141,434739560)

Reply:
>  That means there can only ever be one type per driver instance.

Consider the case where I have authored a library `Generator.Utility.dll` which has an accessible type `SimpleGenerator` that implements `ISourceGenerator`. Two different generators could reference this library and expose that type. 

Hmmm ... Or do we scan the generator assembly directly and look for derivations of `ISourceGenerator`? That wouldn't have this problem but if we had a factory model we would have this issue.

Reply:
We scan for a combo of derived form `ISourceGenerator` and marked with the `SourceGeneratorAttribute`. I'll have to double check, but I believe in your example we'd still only end up with one generator (we use the same mechanism as analyzer discovery, so its what ever that does)

---
In reply to: [434766641](https://github.com/dotnet/roslyn/pull/44768#discussion_r434766641) [](ancestors = 434766641)